core/shared/src/main/scala/fs2/pipe.scala

```diff
@@ -75,6 +75,20 @@ object pipe {
   def delete[F[_],I](p: I => Boolean): Stream[F,I] => Stream[F,I] =
     _ pull { h => h.takeWhile((i:I) => !p(i)).flatMap(_.drop(1)).flatMap(_.echo) }
 
+  /**
+   * Emits only elements that are distinct from their immediate predecessors,
+   * using natural equality for comparison.
+   */
+  def distinctConsecutive[F[_],I]: Stream[F,I] => Stream[F,I] =
+    filterBy2((i1, i2) => i1 != i2)
+
+  /**
+   * Emits only elements that are distinct from their immediate predecessors
+   * according to `f`, using natural equality for comparison.
+   */
+  def distinctConsecutiveBy[F[_],I,I2](f: I => I2): Stream[F,I] => Stream[F,I] =
+    filterBy2((i1, i2) => f(i1) != f(i2))
```

Comment:
Not sure if this matters, but this will call `f` twice on the same element, once when it is being compared to its previous element, and once when being compared with the next element. Might be worth documenting this.

You could also implement `distinctConsecutiveBy` via a `map` (create a pair of `(i, f(i))`), then the `filterBy2` operating on the `f(i)` portion of the pair, then a `map(_._1)` to get back the `i`. Don't think I would do this though. Let users do that before calling the function.